src/juju/types.ts

```diff
@@ -0,0 +1,217 @@
+// See https://github.com/juju/juju/blob/develop/apiserver/params/multiwatcher.go
+// for the Juju types for the AllWatcher responses.
+
+export interface ModelWatcherData {
+  [uuid: string]: ModelData;
+}
+
+export type AllWatcherDelta =
+  | ["action", "change", ActionChangeDelta]
+  | ["application", "change", ApplicationChangeDelta]
+  | ["charm", "change", CharmChangeDelta]
+  | ["unit", "change", UnitChangeDelta]
+  | ["machine", "change", MachineChangeDelta]
+  | ["model", "change", ModelChangeDelta]
+  | ["relation", "change", RelationChangeDelta];
+
+interface ModelData {
+  applications: Applications;
+}
+
+interface Applications {
+  name: string;
+}
+
+// Shared Types
+
+type IPAddress = string;
+type UnitId = string;
+type NumberAsString = string;
+type Life = "alive" | "dead" | "dying";
+type ISO8601Date = string;
+type DeprecatedString = string;
+interface Status {
+  // See https://github.com/juju/juju/blob/develop/core/status/status.go
+  // For the possible status values for `current`.
+  // Possible statuses differ by entity type.
+  current: string;
+  message: string;
+  since: ISO8601Date;
+  version: string;
+  data: { [key: string]: any };
+  err?: string;
+}
+
+// Delta Types
+
+interface ActionChangeDelta {
+  "model-uuid": string;
+  id: NumberAsString;
+  receiver: UnitId;
+  name: string;
+  status: string;
+  message: string;
+  parameters?: { [key: string]: string };
+  results?: {
+    // Known values, add more as known.
+    Code: NumberAsString;
+    Stderr: string;
+    [key: string]: string;
+  };
+  enqueued: ISO8601Date;
+  started: ISO8601Date;
+  completed: ISO8601Date;
+}
+
+interface ApplicationChangeDelta {
+  "workload-version": string;
+  "charm-url": string;
+  "min-units": number;
+  "model-uuid": string;
+  "owner-tag": string;
+  constraints: { [key: string]: string };
+  exposed: boolean;
+  life: Life;
+  name: string;
+  status: WorkloadStatus;
+  subordinate: boolean;
+}
+
+interface CharmChangeDelta {
+  "model-uuid": string;
+  "charm-url": string;
+  "charm-version": string;
+  life: Life;
+  profile: {
+    config?: { [key: string]: string };
+    description?: string;
+    devices?: {
+      [key: string]: { [key: string]: string };
+    };
+  } | null; // lxdprofile
+  config?: { [key: string]: string | boolean };
+}
+
+interface MachineChangeDelta {
+  addresses: AddressData | null;
+  "agent-status": MachineAgentStatus;
+  "container-type": string;
+  "hardware-characteristics": HardwareCharacteristics | undefined;
```

Comment:
Should this use a `?` instead of `| undefined` e.g. `"hardware-characteristics"?: HardwareCharacteristics`?

Reply:
Yep! good catch.